### PR TITLE
Check workitem status, add more verbose logging and handle exceptions more gracefully

### PR DESCRIPTION
**Pull request by karajas**
/cc @mmitche

### CODE REVIEW

File: vars/waitForHelixRuns.groovy [hunk 1]
```diff
@@ -25,15 +25,25 @@ def call (def helixRunsBlob, String prStatusPrefix) {
...
-        assert statusResponse != null
-        assert statusResponse.content != null
-        def statusContent = (new JsonSlurperClassic()).parseText(statusResponse.content)
-        def mcResultsUrl = "https://mc.dot.net/#/user/${getEncodedUrl(statusContent.Creator)}/${getEncodedUrl(statusContent.Source)}/${getEncodedUrl(statusContent.Type)}/${getEncodedUrl(statusContent.Build)}"
+        try {
+            def currentRun = helixRunsBlob[i];
+            def queueId = currentRun['QueueId']
+            def correlationId = currentRun['CorrelationId']
+            def statusUrl = "https://helix.dot.net/api/2017-04-14/jobs/${correlationId}/details"
+            println('Sending request to ' + statusUrl)
```
Comment (mmitche) [inline vars/waitForHelixRuns.groovy]: I believe the url is directly echoed in the httpRequest step below.

File: vars/waitForHelixRuns.groovy [hunk 2]
```diff
@@ -52,109 +62,126 @@ def call (def helixRunsBlob, String prStatusPrefix) {
...
-                def statusContent = (new JsonSlurperClassic()).parseText(statusResponse.content)
+                try{
+                    // Check the state against the Helix API
+                    def statusUrl = "https://helix.dot.net/api/2017-04-14/jobs/${correlationId}/details"
+                    println('Sending request to ' + statusUrl)
+                    def statusResponse = httpRequest statusUrl
+                    assert statusResponse != null
+                    assert statusResponse.content != null
+                    def statusContent = (new JsonSlurperClassic()).parseText(statusResponse.content)
+                    println('Response received from ' + statusUrl + ': ' + statusContent)
```
Comment (mmitche) [inline vars/waitForHelixRuns.groovy]: You can echo the response by adding a parameter to httpRequest: consoleLogResponseBody=true

https://jenkins.io/doc/pipeline/steps/http_request/

File: vars/waitForHelixRuns.groovy [hunk 3]
```diff
@@ -52,109 +62,126 @@ def call (def helixRunsBlob, String prStatusPrefix) {
...
-                    def resultsContent = (new JsonSlurperClassic()).parseText(resultsResponse.content)
+                    def resultValue
+                    def subMessage
+                    // If it's running, grab the current state of results too
+                    if (isRunning || isFinished) {
+                        // Check the results
+                        // We check the results by going to the API aggregating by correlation id
+                        def resultsUrl = "https://helix.dot.net/api/2017-04-14/aggregate/jobs?groupBy=job.name&maxResultSets=1&filter.name=${correlationId}"
+
+                        println('Sending request to ' + resultsUrl)
```
Comment (mmitche) [inline vars/waitForHelixRuns.groovy]: Shouldn't need to echo the request here either

File: vars/waitForHelixRuns.groovy (hunk 1 above)
Comment (karajas) [inline vars/waitForHelixRuns.groovy]: I thought it would be better to track explicit user created echos but I guess I can remove this.

**Comment by karajas** [comment]
PTAL @mmitche

**Comment by karajas** [comment]
@dotnet-bot test CI Tests

**Comment by mmitche** [comment]
@karajas LGTM.  I wonder if we can add a test for this.  Is there a way to send test payloads somewhere?  We could write a little part of the test pipeline which would send them, then call this wait step.

**Comment by karajas** [comment]
We could make a test which basically ran Hello world through Helix and then test this function.

**Comment by mmitche** [comment]
@karajas That would be perfect.  Want to give it a quick try?  See https://github.com/dotnet/dotnet-ci/blob/dev/pipelinesupport/tests/pipeline/pipeline-tests.groovy

File: vars/waitForHelixRuns.groovy [hunk 4]
```diff
@@ -25,15 +25,23 @@ def call (def helixRunsBlob, String prStatusPrefix) {
...
-        def currentRun = helixRunsBlob[i];
-        def queueId = currentRun['QueueId']
-        def correlationId = currentRun['CorrelationId']
-        def statusUrl = "https://helix.dot.net/api/2017-04-14/jobs/${correlationId}/details"
-        def statusResponse = httpRequest statusUrl
-        assert statusResponse != null
-        assert statusResponse.content != null
-        def statusContent = (new JsonSlurperClassic()).parseText(statusResponse.content)
-        def mcResultsUrl = "https://mc.dot.net/#/user/${getEncodedUrl(statusContent.Creator)}/${getEncodedUrl(statusContent.Source)}/${getEncodedUrl(statusContent.Type)}/${getEncodedUrl(statusContent.Build)}"
+        try {
```
Comment (karajas) [inline vars/waitForHelixRuns.groovy]: Does having a try catch here make sense? So that it at leasts attempts to check every correlation id?

Comment (mmitche) [inline vars/waitForHelixRuns.groovy]: Yeah I like that here.

**Comment by karajas** [comment]
I might not be able to add the test right now due to some First Responder work, but will file an separate issue to add this tests to CI tests.

**Comment by karajas** [comment]
@mmitche I'll kick off a corefx run with these changes before merging.

**Comment by karajas** [comment]
https://github.com/dotnet/corefx/pull/20778 seems to indicate the right things are visible in the log https://ci3.dot.net/job/dotnet_corefx/job/master/job/portable-windows-Config_Release+OuterLoop_false_prtest/905/console

**Comment by karajas** [comment]
The test case we add should also have the negative case as well. I'll be sure to include that in the issue.

**Pull request merged**